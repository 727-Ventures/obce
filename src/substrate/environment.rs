use frame_support::dispatch::Weight;
use frame_system::Config as SysConfig;
use pallet_contracts::chain_extension::{
    BufInBufOutState,
    Environment,
    Ext,
    Result,
};
use sp_core::{
    Decode,
    MaxEncodedLen,
};
use sp_std::vec::Vec;

/// Generalized chain extension execution environment.
///
/// Custom [`ChainExtensionEnvironment`] implementations can be used for chain extension testing.
pub trait ChainExtensionEnvironment<E, T> {
    /// Opaque charged weight amount handle.
    type ChargedAmount;

    /// The function id within the `id` passed by a contract.
    ///
    /// It returns the two least significant bytes of the `id` passed by a contract as the other
    /// two bytes represent the chain extension itself (the code which is calling this function).
    fn func_id(&self) -> u16;

    /// The chain extension id within the `id` passed by a contract.
    ///
    /// It returns the two most significant bytes of the `id` passed by a contract which represent
    /// the chain extension itself (the code which is calling this function).
    fn ext_id(&self) -> u16;

    /// The length of the input as passed in as `input_len`.
    ///
    /// A chain extension would use this value to calculate the dynamic part of its
    /// weight. For example a chain extension that calculates the hash of some passed in
    /// bytes would use `in_len` to charge the costs of hashing that amount of bytes.
    /// This also subsumes the act of copying those bytes as a benchmarks measures both.
    fn in_len(&self) -> u32;

    /// Reads `min(max_len, in_len)` from contract memory.
    ///
    /// This does **not** charge any weight. The caller must make sure that the an
    /// appropriate amount of weight is charged **before** reading from contract memory.
    /// The reason for that is that usually the costs for reading data and processing
    /// said data cannot be separated in a benchmark. Therefore a chain extension would
    /// charge the overall costs either using `max_len` (worst case approximation) or using
    /// [`in_len()`](Self::in_len).
    fn read(&self, max_len: u32) -> Result<Vec<u8>>;

    /// Reads `min(buffer.len(), in_len) from contract memory.
    ///
    /// This takes a mutable pointer to a buffer fills it with data and shrinks it to
    /// the size of the actual data. Apart from supporting pre-allocated buffers it is
    /// equivalent to to [`read()`](Self::read).
    fn read_into(&self, buffer: &mut &mut [u8]) -> Result<()>;

    /// Reads and decodes a type with a size fixed at compile time from contract memory.
    ///
    /// This function is secure and recommended for all input types of fixed size
    /// as long as the cost of reading the memory is included in the overall already charged
    /// weight of the chain extension. This should usually be the case when fixed input types
    /// are used.
    fn read_as<U: Decode + MaxEncodedLen>(&mut self) -> Result<U>;

    /// Reads and decodes a type with a dynamic size from contract memory.
    ///
    /// Make sure to include `len` in your weight calculations.
    fn read_as_unbounded<U: Decode>(&mut self, len: u32) -> Result<U>;

    /// Write the supplied buffer to contract memory.
    ///
    /// If the contract supplied buffer is smaller than the passed `buffer` an `Err` is returned.
    /// If `allow_skip` is set to true the contract is allowed to skip the copying of the buffer
    /// by supplying the guard value of `pallet-contracts::SENTINEL` as `out_ptr`. The
    /// `weight_per_byte` is only charged when the write actually happens and is not skipped or
    /// failed due to a too small output buffer.
    fn write(&mut self, buffer: &[u8], allow_skip: bool, weight_per_byte: Option<Weight>) -> Result<()>;

    /// Charge the passed `amount` of weight from the overall limit.
    ///
    /// It returns `Ok` when there the remaining weight budget is larger than the passed
    /// `weight`. It returns `Err` otherwise. In this case the chain extension should
    /// abort the execution and pass through the error.
    ///
    /// # Note
    ///
    /// Weight is synonymous with gas in substrate.
    fn charge_weight(&mut self, amount: Weight) -> Result<Self::ChargedAmount>;

    /// Adjust a previously charged amount down to its actual amount.
    ///
    /// This is when a maximum a priori amount was charged and then should be partially
    /// refunded to match the actual amount.
    fn adjust_weight(&mut self, charged: Self::ChargedAmount, actual_weight: Weight);

    /// Grants access to the execution environment of the current contract call.
    ///
    /// Consult the functions on the returned type before re-implementing those functions.
    fn ext(&mut self) -> &mut E;
}

impl<'a, 'b, E, T> ChainExtensionEnvironment<E, T> for Environment<'a, 'b, E, BufInBufOutState>
where
    T: SysConfig,
    E: Ext<T = T>,
{
    type ChargedAmount = pallet_contracts::chain_extension::ChargedAmount;

    fn func_id(&self) -> u16 {
        Environment::func_id(self)
    }

    fn ext_id(&self) -> u16 {
        Environment::ext_id(self)
    }

    fn in_len(&self) -> u32 {
        Environment::in_len(self)
    }

    fn read(&self, max_len: u32) -> Result<Vec<u8>> {
        Environment::read(self, max_len)
    }

    fn read_into(&self, buffer: &mut &mut [u8]) -> Result<()> {
        Environment::read_into(self, buffer)
    }

    fn read_as<U: Decode + MaxEncodedLen>(&mut self) -> Result<U> {
        Environment::read_as(self)
    }

    fn read_as_unbounded<U: Decode>(&mut self, len: u32) -> Result<U> {
        Environment::read_as_unbounded(self, len)
    }

    fn write(&mut self, buffer: &[u8], allow_skip: bool, weight_per_byte: Option<Weight>) -> Result<()> {
        Environment::write(self, buffer, allow_skip, weight_per_byte)
    }

    fn charge_weight(&mut self, amount: Weight) -> Result<Self::ChargedAmount> {
        Environment::charge_weight(self, amount)
    }

    fn adjust_weight(&mut self, charged: Self::ChargedAmount, actual_weight: Weight) {
        Environment::adjust_weight(self, charged, actual_weight)
    }

    fn ext(&mut self) -> &mut E {
        Environment::ext(self)
    }
}
