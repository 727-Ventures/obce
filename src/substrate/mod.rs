// Copyright (c) 2012-2022 Supercolony
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the"Software"),
// to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

mod environment;
mod is_critical_error;

pub use environment::ChainExtensionEnvironment;
pub use frame_support;
pub use frame_system;
pub use is_critical_error::{
    ToCriticalErr,
    ToCriticalErrFallback,
};
pub use pallet_contracts;
pub use sp_core;
pub use sp_runtime;
pub use sp_std;

use core::marker::PhantomData;

use pallet_contracts::chain_extension::RetVal;
use sp_runtime::DispatchError;

/// Callable chain extension with generalized environment information.
///
/// Unlike [`ChainExtension`](pallet_contracts::chain_extension::ChainExtension), [`CallableChainExtension`]
/// provides capabilities to call chain extensions with generalized [`ChainExtensionEnvironment`],
/// improving chain extension testing capabilities on the Substrate side.
///
/// This trait is automatically implemented on your Substrate chain extension struct
/// with [`#[obce::implementation]`](macro@crate::implementation) expansion.
///
/// # Generics
///
/// This trait mimics the generics that are used with [`#[obce::implementation]`](macro@crate::implementation) macro.
pub trait CallableChainExtension<E, T, Env> {
    /// Call chain extension with the provided [`ChainExtensionEnvironment`] implementation.
    fn call(&mut self, env: Env) -> Result<RetVal, CriticalError>;
}

/// Chain extension context that you can use with your implementations.
///
/// # Generics
///
/// This struct mimics the generics that are used with [`#[obce::implementation]`](macro@crate::implementation) macro.
pub struct ExtensionContext<'a, E, T, Env, Extension>
where
    Env: ChainExtensionEnvironment<E, T>,
{
    /// Chain extension environment.
    pub env: Env,

    /// Custom chain extension storage.
    pub storage: &'a mut Extension,

    pre_charged: Option<Env::ChargedAmount>,

    _ghost: PhantomData<(E, T)>,
}

impl<'a, E, T, Env, Extension> ExtensionContext<'a, E, T, Env, Extension>
where
    Env: ChainExtensionEnvironment<E, T>,
{
    pub fn new(storage: &'a mut Extension, env: Env, pre_charged: Option<Env::ChargedAmount>) -> Self {
        ExtensionContext {
            env,
            storage,
            pre_charged,
            _ghost: PhantomData,
        }
    }

    /// Get the pre-charged amount of weight.
    ///
    /// This method returns [`None`] if there was no pre-charge made,
    /// or if you previously called this method, since there is no copy
    /// of a charged amount involved.
    pub fn pre_charged(&mut self) -> Option<Env::ChargedAmount> {
        self.pre_charged.take()
    }
}

pub type CriticalError = DispatchError;

/// The trait allows filtering error on critical and non-critical errors.
///
/// Critical errors terminate the execution of the chain extension, while
/// non-critical errors are propagated to the caller contract via buffer.
pub trait SupportCriticalError: Sized {
    /// Try to convert an error to a critical one.
    ///
    /// Returns [`Ok`] if the conversion was successful (and the current
    /// error should be qualified as critical), and [`Err`] otherwise.
    fn try_to_critical(self) -> Result<CriticalError, Self>;
}
