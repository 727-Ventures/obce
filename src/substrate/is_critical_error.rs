// Copyright (c) 2012-2022 Supercolony
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the"Software"),
// to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

use crate::substrate::{
    CriticalError,
    SupportCriticalError,
};

pub struct ToCriticalErr<T>(pub T);

impl<T, E> ToCriticalErr<Result<T, E>>
where
    E: SupportCriticalError,
{
    #[inline]
    // We need to allow for dead code at this point because
    // the Rust compiler thinks this function is unused even
    // though it acts as the specialized case for detection.
    #[allow(dead_code)]
    pub fn try_to_critical_error(self) -> Result<Result<T, E>, CriticalError> {
        match self.0 {
            Ok(result) => Ok(Ok(result)),
            Err(error) => {
                match error.try_to_critical() {
                    Ok(critical_error) => Err(critical_error),
                    Err(not_critical) => Ok(Err(not_critical)),
                }
            }
        }
    }
}

pub trait ToCriticalErrFallback<T> {
    fn try_to_critical_error(self) -> Result<T, CriticalError>;
}
impl<T> ToCriticalErrFallback<T> for ToCriticalErr<T> {
    #[inline]
    fn try_to_critical_error(self) -> Result<T, CriticalError> {
        Ok(self.0)
    }
}

#[macro_export]
#[doc(hidden)]
macro_rules! to_critical_error {
    ( $e:expr $(,)? ) => {{
        #[allow(unused_imports)]
        use $crate::substrate::ToCriticalErrFallback;
        $crate::substrate::ToCriticalErr($e).try_to_critical_error()
    }};
}

#[cfg(test)]
mod tests {
    use crate::substrate::{
        CriticalError,
        SupportCriticalError,
    };

    #[test]
    fn to_critical_error_works_if_trait_implemented() {
        #[derive(PartialEq, Eq, Debug)]
        enum Error {
            NonCritical,
            Critical(CriticalError),
        }

        impl SupportCriticalError for Error {
            fn try_to_critical(self) -> Result<CriticalError, Self> {
                match self {
                    Error::Critical(critical) => Ok(critical),
                    _ => Err(self),
                }
            }
        }

        let error: Result<(), _> = Err(Error::Critical(CriticalError::BadOrigin));
        assert_eq!(to_critical_error!(error), Err(CriticalError::BadOrigin));

        let error: Result<(), _> = Err(Error::NonCritical);
        assert_eq!(to_critical_error!(error), Ok(Err(Error::NonCritical)));
    }

    #[test]
    fn to_critical_error_works_if_trait_is_not_implemented() {
        #[derive(PartialEq, Eq, Debug)]
        enum Error {
            NonCritical,
            Critical(CriticalError),
        }

        let error: Result<(), _> = Err(Error::Critical(CriticalError::BadOrigin));
        assert_eq!(
            to_critical_error!(error),
            Ok(Err(Error::Critical(CriticalError::BadOrigin)))
        );

        let error: Result<(), _> = Err(Error::NonCritical);
        assert_eq!(to_critical_error!(error), Ok(Err(Error::NonCritical)));
    }

    #[test]
    fn to_critical_error_works_with_without_result() {
        let result = ();
        assert_eq!(to_critical_error!(result), Ok(()));
    }
}
