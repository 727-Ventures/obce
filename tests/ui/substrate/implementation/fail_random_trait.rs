use obce::substrate::{
    frame_system::Config as SysConfig,
    pallet_contracts::{
        chain_extension::Ext,
        Config as ContractConfig,
    },
    sp_runtime::traits::StaticLookup,
    ExtensionContext
};

pub struct ChainExtension;

pub trait ChainExtensionDefinition {}

#[obce::implementation]
impl<'a, 'b, E, T> ChainExtensionDefinition for ExtensionContext<'a, 'b, E, T, ChainExtension>
where
    T: SysConfig + ContractConfig,
    <<T as SysConfig>::Lookup as StaticLookup>::Source: From<<T as SysConfig>::AccountId>,
    E: Ext<T = T>,
{
}

fn main() {}
