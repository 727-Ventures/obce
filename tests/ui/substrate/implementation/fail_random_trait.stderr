error[E0277]: the trait bound `dyn ChainExtensionDefinition: ExtensionDescription` is not satisfied
  --> tests/ui/substrate/implementation/fail_random_trait.rs:15:1
   |
15 | #[obce::implementation]
   | ^^^^^^^^^^^^^^^^^^^^^^^ the trait `ExtensionDescription` is not implemented for `dyn ChainExtensionDefinition`
   |
   = note: this error originates in the attribute macro `obce::implementation` (in Nightly builds, run with -Z macro-backtrace for more info)
